reference/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -0,0 +1,332 @@
+            } else {
+                // no neighbor
+                strongest_neighbor.get_data()[row] = row;
+            }
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_NON_COMPLEX_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_AMGX_PGM_FIND_STRONGEST_NEIGHBOR);
+
+
+template <typename ValueType, typename IndexType>
+void assign_to_exist_agg(std::shared_ptr<const ReferenceExecutor> exec,
+                         const matrix::Csr<ValueType, IndexType> *weight_mtx,
+                         const matrix::Diagonal<ValueType> *diag,
+                         Array<IndexType> &agg,
+                         Array<IndexType> &intermediate_agg)
+{
+    const auto row_ptrs = weight_mtx->get_const_row_ptrs();
+    const auto col_idxs = weight_mtx->get_const_col_idxs();
+    const auto vals = weight_mtx->get_const_values();
+    const auto agg_const_val = agg.get_const_data();
+    auto agg_val = (intermediate_agg.get_num_elems() > 0)
+                       ? intermediate_agg.get_data()
+                       : agg.get_data();
+    const auto diag_vals = diag->get_const_values();
+    for (IndexType row = 0; row < agg.get_num_elems(); row++) {
+        if (agg_const_val[row] != -1) {
+            continue;
+        }
+        auto max_weight_agg = zero<ValueType>();
+        IndexType strongest_agg = -1;
+        for (auto idx = row_ptrs[row]; idx < row_ptrs[row + 1]; idx++) {
+            auto col = col_idxs[idx];
+            if (col == row) {
+                continue;
+            }
+            auto weight =
+                vals[idx] / max(abs(diag_vals[row]), abs(diag_vals[col]));
+            if (agg_const_val[col] != -1 &&
+                (weight > max_weight_agg ||
+                 (weight == max_weight_agg && col > strongest_agg))) {
+                max_weight_agg = weight;
+                strongest_agg = col;
+            }
+        }
+        if (strongest_agg != -1) {
+            agg_val[row] = agg_const_val[strongest_agg];
+        } else {
+            agg_val[row] = row;
+        }
+    }
+
+    if (intermediate_agg.get_num_elems() > 0) {
+        // Copy the intermediate_agg to agg
+        agg = intermediate_agg;
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_NON_COMPLEX_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_AMGX_PGM_ASSIGN_TO_EXIST_AGG);
+
+
+template <typename ValueType, typename IndexType>
+void amgx_pgm_generate(std::shared_ptr<const ReferenceExecutor> exec,
+                       const matrix::Csr<ValueType, IndexType> *source,
+                       const Array<IndexType> &agg,
+                       matrix::Csr<ValueType, IndexType> *coarse)
+{
+    // agg[i] -> I, agg[j] -> J
+    const auto coarse_nrows = coarse->get_size()[0];
+    const auto source_nrows = source->get_size()[0];
+    const auto source_row_ptrs = source->get_const_row_ptrs();
+    const auto source_col_idxs = source->get_const_col_idxs();
+    const auto source_vals = source->get_const_values();
+    vector<map<IndexType, ValueType>> row_list(
```

Comment:
If these are `gko` objects I would maybe prefer to use `gko::` here

Reply:
We don't have these? They should be `std` I believe.

Reply:
These are typedefs using ExecutorAllocator

Reply:
Yes, it is in `gko` namespace

Reply:
Thanks, I did not remember we had these.